static/src/javascripts/projects/common/modules/commercial/slice-adverts.js

```diff
@@ -1,120 +1,118 @@
-                fabricAdSlot.addClass('ad-slot--mobile');
-                fabricAdSlot.insertAfter(containers[0]);
-            }
+    function init() {
+        if (!commercialFeatures.sliceAdverts) {
+            return false;
+        }
 
-            // skip the initial containers if we've already added an advert to the first
-            index+= addFabricAd ? (1 + containerGap) : 0;
+        var fabricAdSlot, adSlots;
+        var prefs          = userPrefs.get('container-states') || {};
+        var isMobile       = detect.getBreakpoint() === 'mobile';
+        var isNetworkFront = ['uk', 'us', 'au'].indexOf(config.page.pageId) !== -1;
+        var hasFabricAd    = (config.page.isFront && config.switches.fabricAdverts && detect.isBreakpoint({max : 'phablet'}));
 
+        adSlots = qwery(containerSelector)
+            // get all ad slices
+            .map(function (container) {
+                return {
+                    container: container,
+                    adSlice: container.querySelector(sliceSelector)
+                };
+            })
             // pull out ad slices which have at least x containers between them
-            while (index < containers.length) {
-                container    = containers[index];
-                containerId  = bonzo(container).data('id');
-                $adSlice     = $(opts.sliceSelector, container);
-                // don't display ad in the first container on the network fronts
-                isFrontFirst = contains(['uk', 'us', 'au'], config.page.pageId) && index === 0;
-
+            .filter(function (item, index) {
                 if (config.page.showMpuInAllContainers) {
-                    adSlices.push($adSlice.first());
-                    index++;
-                } else {
-                    if ($adSlice.length && !isFrontFirst && (!prefs || prefs[containerId] !== 'closed')) {
-                        adSlices.push($adSlice.first());
-                        index += (containerGap + 1);
-                    } else {
-                        index++;
-                    }
+                    return true;
                 }
-            }
 
-            return Promise.all(chain(adSlices).slice(0, maxAdsToShow).and(map, function ($adSlice, index) {
-                    // When we are inside the AB test we are adding inline1 manually so index needs to start from 2.
-                    var inlineIndexOffset = (config.tests.cmTopBannerPosition) ? 2 : 1;
+                // Slots are inserted every containerGap + 1 slice
+                if (!item.adSlice || index % (containerGap + 1) !== 0) {
+                    return false;
+                }
+
+                var containerId  = item.container.getAttribute('data-id');
+                var isContainerClosed = prefs[containerId] === 'closed';
+                var isFrontFirst = isNetworkFront && index === 0;
+                return !(isFrontFirst || isContainerClosed);
```

Comment:
I know there's a little more overhead to running more filter functions, but would these benefit from being separated into their own distinct filters?